compiler/rustc_mir_transform/src/instcombine.rs

```diff
@@ -0,0 +1,405 @@
+}
+
+struct Context<'tcx, 'a> {
+    tcx: TyCtxt<'tcx>,
+    local_decls: &'a LocalDecls<'tcx>,
+    param_env: ParamEnv<'tcx>,
+}
+
+// Combine:
+// _2 = &_1;
+// _3 = *_2;
+// Into:
+// _3 = _1;
+fn combine_ref_deref<'tcx, 'a>(
+    cx: &Context<'tcx, 'a>,
+    _temp_place: &Place<'tcx>,
+    temp_rvalue: &Rvalue<'tcx>,
+    final_place: &Place<'tcx>,
+    final_rvalue: &Rvalue<'tcx>,
+) -> Option<StatementKind<'tcx>> {
+    let Rvalue::Ref(_, _, first_place) = temp_rvalue else { return None; };
+    let Rvalue::Use(final_operand) = final_rvalue else { return None; };
+
+    let Some(second_place) = final_operand.place() else { return None; };
+    if second_place.projection.get(0) != Some(&ProjectionElem::Deref) {
+        return None;
+    }
+
+    let new_place = first_place.project_deeper(&second_place.projection[1..], cx.tcx);
+
+    if new_place == *final_place {
+        Some(StatementKind::Nop)
+    } else {
+        Some(StatementKind::Assign(Box::new((*final_place, Rvalue::Use(Operand::Copy(new_place))))))
+    }
+}
+
+// Combine:
+// _2 = &mut _1;
+// _3 = &raw mut *_2;
+// Into:
+// _3 = &raw mut _1;
+fn combine_ref_addressof<'tcx, 'a>(
+    cx: &Context<'tcx, 'a>,
+    _temp_place: &Place<'tcx>,
+    temp_rvalue: &Rvalue<'tcx>,
+    final_place: &Place<'tcx>,
+    final_rvalue: &Rvalue<'tcx>,
+) -> Option<StatementKind<'tcx>> {
+    let Rvalue::Ref(_, _, first_place) = temp_rvalue else { return None; };
+    let Rvalue::AddressOf(mutability, second_place) = final_rvalue else { return None; };
+
+    if second_place.projection.get(0) != Some(&ProjectionElem::Deref) {
+        return None;
+    }
+
+    let new_place = first_place.project_deeper(&second_place.projection[1..], cx.tcx);
+
+    Some(StatementKind::Assign(Box::new((*final_place, Rvalue::AddressOf(*mutability, new_place)))))
+}
+
+// Combine:
+// _2 = _1.a;
+// _3 = _2.b;
+// Into:
+// _3 = _1.a.b;
+fn combine_place_projections<'tcx, 'a>(
+    cx: &Context<'tcx, 'a>,
+    temp_place: &Place<'tcx>,
+    temp_rvalue: &Rvalue<'tcx>,
+    final_place: &Place<'tcx>,
+    final_rvalue: &Rvalue<'tcx>,
+) -> Option<StatementKind<'tcx>> {
+    let Rvalue::Use(temp_operand) = temp_rvalue else { return None; };
+    let Rvalue::Use(final_operand) = final_rvalue else { return None; };
+
+    // Both the operands needs to be places (is it even possible for one to be a Constant?)
+    let first_place = temp_operand.place()?;
+    let second_place = final_operand.place()?;
+
+    // If we are assigning into a place expression, that would be something like
+    // _3.b = _2.b;
+    // Which is complicated. Just don't optimize that at all for now.
+    if !temp_place.projection.is_empty() {
+        return None;
+    }
+
+    // Derefs must come first if at all
+    // If merging these assignments would break that rule, bummer. Bail.
+    if first_place.projection.len() > 0
+        && second_place.projection.get(0) == Some(&ProjectionElem::Deref)
+    {
+        return None;
+    }
```

Comment:
Likewise, should we check that `temp_place.local == second_place.local`?
For instance:
```
_2 = _1.a
_3 = _4[_2].b
```